### DISCUSSION TITLE
How can we know whether solver was able to solve the problem or not?

### DISCUSSION THREAD

**Discussion by prateekkrjain** [qpsolvers/qpsolvers#35] [Q&A] [answered]
Hi,

In CVXOPT we get parameter - 'STATUS', but solve_qp is only returning the 'x'. So, how can we know if the solution is optimal or not? Is there any way to check that?

Thanks,

**Comment by stephane-caron** [comment]
**Edit:** _now there is! 🎉 See below_. 👇

Previously, the way to go was to check whether the return value of ``solve_qp`` was `None`:

- If the return value is `None`, the solver was not able to solve the problem.
- If the return value is an array, the solver found a solution.

In some cases knowing that the solver found a solution is not fully satisfactory. For instance, it could have settled on the best value so far but with no guarantee to be less than some threshold away from the true optimum (**Edit:** _those guarantees are the residuals now checked by ``Solution.is_optimal`` below_.).

There's no mechanism to relay other outputs, e.g. dual multipliers or solver status, (**Edit:** _now we can get both dual multipliers and solver status_.), but you can still do it on a per-solver basis. For example, if you know you want to use CVXOPT, then adapt the `cvxopt_solve_qp` function to return the full solution and read 'STATUS' from there. Or if you test all solvers via qpsolvers and find that e.g. OSQP performs better, then adapt the `osqp_solve_qp` function to your use case so that it returns `res.info.status_val` as well.

In the future the library will offer a way to get those extra outputs when they are provided. For example, an additional `extra_outputs` kwarg for `solve_qp` defaulting to `None` (no action) and filled with status, dual multipliers and so forth when they are available.

**Comment by stephane-caron** [comment]
### Extra solver outputs

> In the future the library will offer a way to get those extra outputs when they are provided.

And the future is here! :wink: It is now possible to retrieve extra solver output, including status flags, in the ``Solution.extras`` dictionary. For instance, assuming we run example [`05_dual_multipliers.py`](https://github.com/stephane-caron/qpsolvers/blob/7e3223c49e76478f21a9a2cd7b02f89fa9330509/examples/05_dual_multipliers.py) with CVXOPT, we can check:

```python
In [7]: solution.extras["status"]                                                                                                                  
Out[7]: 'optimal'
```

Extra solver outputs are specific though. Some solvers have detailed return information, and some don't. Fortunately there is now also a solver-independent way to check a solution: see *Checking whether a solution is optimal or not* below.

> So, how can we know if the solution is optimal or not? Is there any way to check that?

Back to your initial question in the reply below :wink:

**Comment by stephane-caron** [comment, marked as answer]
We can now check the full [optimality conditions](https://scaron.info/blog/optimality-conditions-and-numerical-tolerances-in-qp-solvers.html) on a given solution, *i.e.* a boolean that is ``True`` if and only if the solution is optimal. Here is how to do it:

```python
In [8]: solution.is_optimal(eps_abs=1e-5)                                                                                                          
Out[8]: True
```

Note that there is a parameter $\epsilon_{abs}$ representing the tolerance of the check. You can think of it as the tolerance on equality or inequality constraint violations (although it is also used in some other checks on the dual). For instance, if it is OK for your application that the solution violates inequality constraints $G x \leq h$ by $10^{-5}$ on each coordinate, then set $\epsilon_{abs} = 10^{-5}$.
